gdal/alg/viewshed.cpp

```diff
@@ -0,0 +1,612 @@
+ *
+ * @param pProgressArg The callback data for the pfnProgress function.
+ *
+ * @return CE_None on success or CE_Failure if an error occurs.
+ */
+
+#define EARTH_DIAMETER 12741994.0;
+
+#ifndef round
+inline double round( double d )
+{
+    return floor( d + 0.5 );
+}
+#endif
+
+
+void SetVisibility(int iPixel, double dfZ, double dfZTarget, double* padfZVal,
+    GByte* pabyResult, GByte byVisibleVal, GByte byInvisibleVal)
+{
+    if (padfZVal[iPixel] + dfZTarget < dfZ)
+    {
+        padfZVal[iPixel] = dfZ;
+        pabyResult[iPixel] = byInvisibleVal;
+    }
+    else
+        pabyResult[iPixel] = byVisibleVal;
+}
+
+int AdjustHeightInRange(double* adfGeoTransform, int iPixel, int iLine, double& dfHeight, double dfDistance2, double dfCurvCoeff)
+{  
+    if (dfDistance2 <= 0 && dfCurvCoeff == 0)
+        return TRUE;
+
+    double dfX = adfGeoTransform[1] * iPixel + adfGeoTransform[2] * iLine;
+    double dfY = adfGeoTransform[4] * iPixel + adfGeoTransform[5] * iLine;
+    double dfR2 = dfX * dfX + dfY * dfY;
+
+    /* calc adjustment */
+    if (dfCurvCoeff != 0)
+        dfHeight -= dfCurvCoeff * dfR2 / EARTH_DIAMETER;
+    
+    if (dfDistance2 > 0 && dfR2 > dfDistance2)
+        return FALSE;
+
+    return TRUE;
+}
+
+double CalcHeightLine(int i, double Za, double Zo)
+{
+    if (i == 1)
+        return Za;
+    else
+        return (Za - Zo) / (i - 1) + Za;
+}
+
+double CalcHeightDiagonal(int i, int j, double Za, double Zb, double Zo)
+{
+    return ((Za - Zo) * i + (Zb - Zo) * j) / (i + j - 1) + Zo;
+}
+
+double CalcHeightEdge(int i, int j, double Za, double Zb, double Zo)
+{
+    if (i == j)
+        return CalcHeightLine(i, Za, Zo);
+    else
+        return ((Za - Zo) * i + (Zb - Zo) * (j - i)) / (j - 1) + Zo;
+}
+
+double CalcHeight(double dfZ, double dfZ2, GDALViewshedMode eMode)
+{
+    if (eMode == GVM_Edge)
+        return dfZ2;
+    else if (eMode == GVM_Max)
+        return std::max(dfZ, dfZ2);
+    else if (eMode == GVM_Min)
+        return std::min(dfZ, dfZ2);
+    else
+        return dfZ;
+}
+
+CPLErr GDALViewshedGenerate(GDALRasterBandH hBand, const char* pszTargetRasterName,
+                    double dfObserverX, double dfObserverY, double dfObserverHeight,
+                    double dfTargetHeight, double dfVisibleVal, double dfInvisibleVal,
+                    double dfOutOfRangeVal, double dfNoDataVal, double dfCurvCoeff,
+                    GDALViewshedMode eMode, double dfMaxDistance,
+                    GDALProgressFunc pfnProgress, void *pProgressArg)
+
+{
+    VALIDATE_POINTER1( hBand, "GDALViewshedGenerate", CE_Failure );
+
+    if( pfnProgress == nullptr )
+        pfnProgress = GDALDummyProgress;
+
+    if( !pfnProgress( 0.0, "", pProgressArg ) )
+    {
+        CPLError( CE_Failure, CPLE_UserInterrupt, "User terminated" );
+        return CE_Failure;
+    }
+
+    GByte byNoDataVal = dfNoDataVal >= 0 ? static_cast<GByte>(dfNoDataVal) : 0;
```

Comment:
upper bound should probably be tested as well.  And one could also test CPLIsNan() for pedanticity...